### DISCUSSION TITLE
How to Reuse PDF.js from ngx-extended-pdf-viewer

### DISCUSSION THREAD

**Discussion by 94ostry** [stephanrauh/ngx-extended-pdf-viewer#2695] [Q&A]
Hi everyone,

I'm currently using ngx-extended-pdf-viewer in my Angular project and it works great for most of my use cases. However, I have a specific requirement where I need to perform custom rendering of a PDF document onto a <canvas> element. Instead of adding a new dependency like pdfjs-dist, I’d like to reuse the pdf.js version already bundled with ngx-extended-pdf-viewer.

I found that the pdf.js module is available under the path:
`/ngx-extended-pdf-viewer/assets/pdf.sandbox-4.7.701.mjs`

I tried dynamically importing this module and using it for rendering. Here's my implementation:

```ts
import { Component, AfterViewInit, ViewChild, ElementRef } from '@angular/core';

@Component({
  selector: 'app-pdf-simple-renderer',
  template: `<canvas #pdfCanvas></canvas>`,
  styles: [],
})
export class PdfSimpleRendererComponent implements AfterViewInit {
  pdfPath = 'path-to-your-pdf-file.pdf'; // Path to the PDF file

  async ngAfterViewInit() {
    // Dynamically import pdf.js from ngx-extended-pdf-viewer
    const pdfjs = await import('/ngx-extended-pdf-viewer/assets/pdf.sandbox-4.7.701.mjs');

    // Set up the PDF.js worker
    pdfjs.GlobalWorkerOptions.workerSrc = '/ngx-extended-pdf-viewer/assets/pdf.worker.sandbox-4.7.701.js';

    // Load the PDF document
    const pdfDocument = await pdfjs.getDocument(this.pdfPath).promise;

    // Get the first page of the PDF
    const page = await pdfDocument.getPage(1);

    // Set up the canvas and render the page
    const viewport = page.getViewport({ scale: 1.5 }); // Adjust scale as needed
    const canvas = this.pdfCanvas.nativeElement;
    const context = canvas.getContext('2d')!;
    canvas.width = viewport.width;
    canvas.height = viewport.height;

    const renderContext = {
      canvasContext: context,
      viewport: viewport,
    };

    await page.render(renderContext).promise;
    console.log('PDF rendered successfully!');
  }
}
```

Unfortunately that approach is failing :/ Angular is not able to find pdf.sandbox module.

**Comment by stephanrauh** [comment]
Well, actually, the pdf rendering engine is in the file `pdf-x.x.x.min.mjs` or - since version 21 or so - in `viewer-x.x.x.min.mjs`. The sandbox file is a sandbox for executing JavaScript safely.

You'll also want to double-check the URL. I guess you've found the file in the node_modules folder (e.g. `node_modules/ngx-extended-pdf-viewer/assets/pdf.worker.sandbox-4.7.701.js`), and you expect it to be available unter the same path. However, when you install ngx-extended-pdf-viewer, you add an entry to the `angular.json` that copies these files to the assets folder. Chances are you can load the file from `/assets/pdf.worker.sandbox-4.7.701.mjs`.

Finally, double-check the suffix of the file. It's *.mjs, not *.js.

**Reply by 94ostry** [reply]
Hey, thanks for the tips. Unfortunately, TS keeps throwing an error:

`Cannot find module '/assets/viewer-4.7.701.min.mjs' or its corresponding type declarations.`

So far I tried to modified some value is `tsconfig.json`.
1. added "allowSyntheticDefaultImports": true,

2. declared `src\types\viewer.d.ts` module :
```
declare module '/assets/viewer-4.7.701.min.mjs' {
    export = PDFJS;
}
```

so far without success

**Reply by 94ostry** [reply]
I wonder if adding `pdfjs-dist` dependency won't be the fastest solution. There won't be a place where the `ng-extended-pdf-viewer` component will be displayed on par with plain `pdf.js`, so in theory it should be OK :D

**Reply by stephanrauh** [reply]
Yes, that should work. As far as I know, pdf.js has mostly stopped polluting the global name space, so running pdf.js and my fork of pdf.js in parallel should work.
